iterreg/sparse/solvers.py

```diff
@@ -254,3 +254,42 @@ def fista(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, f_store=10,
             if verbose:
                 print(t, E[t // f_store])
     return w, all_w, E
+
+
+@njit
+def adaptive(X, y, alpha, prox=shrink, pen=ell1, max_iter=1_000, n_adapt=2,
+             f_store=10, w_init=None, verbose=False):
+    """Coordinate descent for the Tikhonov problem."""
+    p = X.shape[1]
+    X = np.asfortranarray(X)
+    lc = np.zeros(p)
+    for j in range(p):
+        lc[j] = norm(X[:, j]) ** 2
+
+    E = np.zeros(max_iter * n_adapt // f_store)
+    adapt_coefs = np.ones(p)
+
+    if w_init is None:
+        w = np.zeros(p)
+        R = y.copy().astype(np.float64)
+    else:
+        w = w_init.copy()
+        R = y - X @ w
+
+    for k in range(n_adapt):
+        for t in range(max_iter):
+            for j in range(p):
+                old = w[j]
+                w[j] = prox(old + X[:, j].dot(R) / lc[j],
+                            alpha*adapt_coefs[j] / lc[j], lc[j])
+                if w[j] != old:
+                    R += ((old - w[j])) * X[:, j]
+            if t % f_store == 0:
+                penalties = [pen(w[j], alpha*adapt_coefs[j]) for j in range(p)]
```

Comment:
need to calrify a bit which penalty it is here : the penalty of the surrogate objective (the reweighted lasso) or the original one ? if it is the original, there should not be weights involved

Reply:
Here it's the reweighted

Reply:
but in that case it should be directly the L1 penalty, not pen. you replace your non convex penalty by a weighted L1